code/modules/urist/modules/GOAI/integrations/base_commander/mob_commander/obstacle_handlers/airlocks.dm

```diff
@@ -0,0 +1,287 @@
+			15 + rand() - (pawn ? get_dist(A, pawn) : 0),
+			1,
+			FALSE,
+			action_args
+		)
+	//Prying unpowered airlocks is an /alternative/ way of opening them, not a prerequisite step. We need an else here
+	else
+		action_key = "[NEED_OBSTACLE_OPEN(A)] for [waypoint]"
+
+		open_door_preconds[action_key] = -TRUE
+		src.SetState(action_key, FALSE)	//Workaround
+
+		var/list/open_door_effects = list()
+		open_door_effects[action_key] = TRUE
+
+		var/list/action_args = list()
+		action_args["obstruction"] = A
+
+		AddAction(
+			action_key,
+			open_door_preconds,
+			open_door_effects,
+			/datum/goai/mob_commander/proc/HandleAirlockOpen,
+			10 + rand() - (pawn ? get_dist(A, pawn) : 0),
+			1,
+			FALSE,
+			action_args
+		)
+
+	return action_key
+
+
+/datum/goai/mob_commander/proc/HandleAirlockOpen(var/datum/ActionTracker/tracker, var/obj/machinery/door/airlock/obstruction)
+	var/mob/pawn = src.GetPawn()
+	if (!pawn || !istype(pawn))
+		return
+
+	if (!tracker)
+		return
+
+	var/walk_dist = (tracker?.BBSetDefault("StartDist", (ManhattanDistance(get_turf(pawn), obstruction) || 0)) || 0)
+
+	if(tracker.IsOlderThan(src.ai_tick_delay * (10 + walk_dist)))
+		if(src.brain.GetMemoryValue(MEM_OBSTRUCTION, null) == obstruction)
+			src.brain.DropMemory(MEM_OBSTRUCTION)
+		tracker.SetFailed()
+		return
+
+	//If the mob can see the airlock, check if its bolt lights are on
+	if(obstruction in src.brain.perceptions[SENSE_SIGHT_CURR])
+		if(obstruction.locked && obstruction.lights)
+			tracker.SetFailed()
+			src.brain.SetMemory(MEM_OBJ_LOCKED(obstruction), TRUE, 5 MINUTES)
+			if(src.brain.GetMemoryValue(MEM_OBSTRUCTION, null) == obstruction)
+				src.brain.DropMemory(MEM_OBSTRUCTION)
+			return
+		if(obstruction.p_open)
+			tracker.SetFailed()
+			src.brain.SetMemory(MEM_OBJ_PANELOPEN(obstruction), TRUE, 5 MINUTES)
+			if(src.brain.GetMemoryValue(MEM_OBSTRUCTION, null) == obstruction)
+				src.brain.DropMemory(MEM_OBSTRUCTION)
+			return
+
+	var/turf/obs_turf = get_turf(obstruction)
+	var/dist_to_obs = ChebyshevDistance(get_turf(pawn), obs_turf)
+
+	if(dist_to_obs < 2 && obstruction.density)
+		//The mob is next to the airlock. Poke it and see what it learns
+		if(obstruction.locked || !obstruction.arePowerSystemsOn())
+			//Airlock won't open (bolted/unpowered). Add it to the memory and fail the tracker.
+			obstruction.attack_hand(pawn)
+			tracker.SetFailed()
+			src.brain.SetMemory(obstruction.locked ? MEM_OBJ_LOCKED(obstruction) : MEM_OBJ_NOPOWER(obstruction), TRUE, 5 MINUTES)
+			if(src.brain.GetMemoryValue(MEM_OBSTRUCTION, null) == obstruction)
+				src.brain.DropMemory(MEM_OBSTRUCTION)
+			return
+		if(!obstruction.allowed(pawn))
+			//No access to the airlock. Add it memory and fail the tracker. Still poke it though for that sweet sweet denied buzz
+			obstruction.attack_hand(pawn)
+			tracker.SetFailed()
+			src.brain.SetMemory(MEM_OBJ_NOACCESS(obstruction), TRUE) //Come back to this and change the ttl if changing IDs/access becomes a thing
```

Comment:
Should probably still use MEM_TIME_LONGTERM rather than the default infinity

Reply:
At the moment we don't have plans to handle the pawn's access changing, so I don't see why we shouldn't store it forever? The access won't change, so it'll always fail here.

Reply:
I guess it would be nicer to leave the door (pun not intended) open for a periodic recheck. Not a big deal though.